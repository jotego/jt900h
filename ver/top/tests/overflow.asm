    ; overflow tests
    maxmode on
    relaxed on
    org 0
    ld a,0xbf    ; common header

    ; 8-bit operands
    ; no overflow add
    ld a,0x40
    add a,0x3f  ; a=0x7f
    jp ov,bad_end
    cp a,0x7f
    jp ne,bad_end
    or xiy,1

    ; overflow add
    ld w,0x7f
    add w,1 ; w=0x80=-128
    jp nov,bad_end
    or xiy,2

    ; no overflow sub
    ld b,0xff
    sub b,1      ; b=0xfe=-2
    jp ov,bad_end
    or xiy,4

    ; overflow sub
    ld c,0x80
    sub c,1      ; c=0x7f=127
    jp nov,bad_end
    or xiy,8

    ; 16-bit operands
    incf
    ld wa,0x7ffe
    ld hl,1
    add wa,hl
    jp ov,bad_end
    or xiy,0x10

    ld bc,0x8000
    sub bc,hl
    jp nov,bad_end
    or xiy,0x20

    ; 32-bit operands
    incf
    ld xwa,0x7ffffffe
    ld xhl,1
    add xwa,xhl
    jp ov,bad_end
    or xiy,0x40

    ; the assembler introduces an unknown (for now)
    ; ed26 instruction after the ld xbc, and ruins
    ; the test, so I comment it out for now
    ; ld xbc,0xffff8000
    ; sub xbc,xhl
    ; jp nov,bad_end
    ; or xiy,0x80

end_loop:
    ldf 0
    ld hl,0xbabe
    ld (0xffff),0xff
    jp end_loop
bad_end:
    ld hl,0xdead
    ld (0xffff),0xff
    jp bad_end
data:
    dw 0xcafe,0xbeef,0xffff,0xeeee,0xcccc
    end