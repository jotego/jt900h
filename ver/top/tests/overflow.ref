0000807f - 00007fff - 7fffffff - 00000000
0000fe7f - 00007fff - 00000000 - 00000000
00000000 - 00000000 - 00000000 - 00000000
0000babe - 00000001 - 00000001 - 00000000
-- index registers --
00000000 - 0027ed7f - 00000000 - 00000100
SR = f800
