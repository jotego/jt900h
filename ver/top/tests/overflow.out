0000807f - 00007fff - 7fffffff - 00000000
0000fe7f - 00007fff - 00000000 - 00000000
00000000 - 00000000 - 00000000 - 00000000
0000babe - 00000001 - 00000001 - 00000000
-- index registers --
00000000 - 0000007f - 00000000 - 00000000
SR = 00f8
